Handler return is not type-checked.

When using TypeScript, the return of a handler is not being validated by the type checker.

**To Reproduce**

Add this file

```typescript
import type {APIGatewayProxyEvent, APIGatewayProxyResult} from "aws-lambda";
import middy from "@middy/core";


export const handler = middy<APIGatewayProxyEvent, APIGatewayProxyResult>()
    .handler(async (req, context) => {
        return 123
    })
```

- Run the type checker.
- Observe that no error is detected.

**Expected behaviour**
The type checker should have throw an error since `123` does not match the `APIGatewayProxyResult`.

**Note:** I know that this can be workaround by manually adding the return type to the handler function, but that's more error prone.

**Environment (please complete the following information):**

- Node.js: 20
- Middy: 5.4.5

This seems to be a regression with Middy 5. Middy 4 is able to correctly validate the type:

```
error TS2345: Argument of type '(req: APIGatewayProxyEvent, context: Context) => Promise<number>' is not assignable to parameter of type 'MiddyInputHandler<APIGatewayProxyEvent, APIGatewayProxyResult, Context>'.
  Type 'Promise<number>' is not assignable to type 'void | APIGatewayProxyResult | Promise<APIGatewayProxyResult>'.
    Type 'Promise<number>' is not assignable to type 'Promise<APIGatewayProxyResult>'.
      Type 'number' is not assignable to type 'APIGatewayProxyResult'.

8     .handler(async (req, context) => {
               ~~~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/packages/core/index.d.ts
+++ b/packages/core/index.d.ts
@@ -99,7 +99,7 @@ export interface MiddyfiedHandler<
   handler: <TInputHandlerEventProps = TEvent, TInputHandlerResultProps = TResult>(
     handler: MiddlewareHandler<
     LambdaHandler<TInputHandlerEventProps, TInputHandlerResultProps>,
-    TContext
+    TContext, TResult
     >
   ) => MiddyfiedHandler<TInputHandlerEventProps, TInputHandlerResultProps, TErr, TContext, TInternal>
 }
@@ -150,8 +150,9 @@ infer TMiddlewareInternal
 
 declare type MiddlewareHandler<
   THandler extends LambdaHandler<any, any>,
-  TContext extends LambdaContext = LambdaContext
-> = THandler extends LambdaHandler<infer TEvent, infer TResult> // always true
+  TContext extends LambdaContext = LambdaContext,
+  TResult = any
+> = THandler extends LambdaHandler<infer TEvent, TResult> // always true
   ? MiddyInputHandler<TEvent, TResult, TContext>
   : never
 
@@ -169,7 +170,7 @@ declare function middy<
 > (
   handler?:
   | LambdaHandler<TEvent, TResult>
-  | MiddlewareHandler<LambdaHandler<TEvent, TResult>, TContext>
+  | MiddlewareHandler<LambdaHandler<TEvent, TResult>, TContext, TResult>
   | PluginObject,
   plugin?: PluginObject
 ): MiddyfiedHandler<TEvent, TResult, TErr, TContext, TInternal>

--- a/packages/core/index.test-d.ts
+++ b/packages/core/index.test-d.ts
@@ -408,4 +408,19 @@ const s3Handler = async (event: S3Event): Promise<void> => {
 }
 
 const handler1182 = middy().handler(s3Handler)
-expectType<MiddyfiedHandler<S3Event, void, Error, Context, {}>>(handler1182)
+expectType<MiddyfiedHandler<S3Event, any, Error, Context, {}>>(handler1182)
+
+//  Issue #1228 Correct return type
+const numberHandler = middy<APIGatewayProxyEvent, number>()
+  .handler(async (event) => {
+    return 42 // Correct return type, should pass type checking
+  })
+expectType<middy.MiddyfiedHandler<APIGatewayProxyEvent, number>>(numberHandler)
+
+//  Issue #1228 Incorrect return type
+const invalidNumberHandler = middy<APIGatewayProxyEvent, number>()
+  // @ts-expect-error
+  .handler(async () => {
+    return 'not a number'
+  })
+expectType<middy.MiddyfiedHandler<APIGatewayProxyEvent, number>>(invalidNumberHandler)

